File: lib/data/models/video_model.dart
class VideoModel {
  final String id;
  final String url;
  final String youtubeId;
  final String? title;
  final String? description;
  final AddedByUser? addedBy;
  final String createdAt;

  VideoModel({
    required this.id,
    required this.url,
    required this.youtubeId,
    this.title,
    this.description,
    this.addedBy,
    required this.createdAt,
  });

  factory VideoModel.fromJson(Map<String, dynamic> json) {
    return VideoModel(
      id: json['id'] ?? json['_id'] ?? '',
      url: json['url'] ?? '',
      youtubeId: json['youtubeId'] ?? '',
      title: json['title'],
      description: json['description'],
      addedBy: json['addedBy'] != null ? AddedByUser.fromJson(json['addedBy']) : null,
      createdAt: json['createdAt'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'url': url,
    'youtubeId': youtubeId,
    'title': title,
    'description': description,
    'addedBy': addedBy?.toJson(),
    'createdAt': createdAt,
  };
}

class AddedByUser {
  final String id;
  final String name;

  AddedByUser({required this.id, required this.name});

  factory AddedByUser.fromJson(Map<String, dynamic> json) {
    return AddedByUser(
      id: json['id'] ?? json['_id'] ?? '',
      name: json['name'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
  };
}

File: arfixbackend/server/src/models/video.model.ts
import mongoose, { Schema, Document, Model } from "mongoose";

export interface IVideo extends Document {
  url: string;
  youtubeId: string;
  title?: string;
  description?: string;
  addedBy?: mongoose.Types.ObjectId;
  createdAt: Date;
  updatedAt: Date;
}

const videoSchema = new Schema<IVideo>(
  {
    url: {
      type: String,
      required: [true, "URL is required"],
      unique: true,
      trim: true,
    },
    youtubeId: {
      type: String,
      required: [true, "YouTube ID is required"],
      trim: true,
    },
    title: {
      type: String,
      trim: true,
      default: null,
    },
    description: {
      type: String,
      trim: true,
      default: null,
    },
    addedBy: {
      type: Schema.Types.ObjectId,
      ref: "User",
      default: null,
    },
  },
  {
    timestamps: true,
    versionKey: false,
  }
);

// Format output
videoSchema.set("toJSON", {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  transform: (_doc: unknown, ret: any) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
    ret.id = ret._id;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    delete ret._id;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    delete ret.__v;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return ret;
  },
});

const Video: Model<IVideo> = mongoose.model<IVideo>("Video", videoSchema);
export default Video;
